### ISSUE TITLE
NPE when running CS with cache with Maven plugin using NewlineAtEndOfFile

### ISSUE THREAD

**Issue by WonderCsabo** [checkstyle/checkstyle#2218]
I am not sure this is the problem of the Maven plugin or CS itself. 

I am seeing this stacktrace after executing `mvn checkstyle:check` twice:

```
Caused by: java.lang.NullPointerException
        at com.puppycrawl.tools.checkstyle.checks.SuppressWarningsHolder.isSuppressed(SuppressWarningsHolder.java:151)
        at com.puppycrawl.tools.checkstyle.filters.SuppressWarningsFilter.accept(SuppressWarningsFilter.java:37)
        at com.puppycrawl.tools.checkstyle.api.FilterSet.accept(FilterSet.java:88)
        at com.puppycrawl.tools.checkstyle.Checker.fireErrors(Checker.java:360)
        at com.puppycrawl.tools.checkstyle.Checker.process(Checker.java:275)
        at org.apache.maven.plugin.checkstyle.exec.DefaultCheckstyleExecutor.executeCheckstyle(DefaultCheckstyleExecutor.java:252)
        at org.apache.maven.plugin.checkstyle.CheckstyleViolationCheckMojo.execute(CheckstyleViolationCheckMojo.java:538)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
```

This only happens when the `checkstyle-cachefile` file is present, from a previous execution. If i delete that file, i do not see this problem.
Also if i remove the `NewlineAtEndOfFile` check, this crash does not happen.

I created a minimal example [project](https://github.com/WonderCsabo/CheckstyleIssue) which reproduces the issue.

**Comment by mkordas** [comment]
@WonderCsabo 
Thanks for the reproducer, it really helped a lot and speeded up investigation. Here is explanation what happends:

* List containing all suppresions is initialized only when tree analysis begins at [SuppressWarningsHolder.java:186](https://github.com/checkstyle/checkstyle/blob/master/src/main/java/com/puppycrawl/tools/checkstyle/checks/SuppressWarningsHolder.java#L186)
* On the second run there is cache file entry saying that for TreeWalker there were no issues on previous run. File was not modified, configuration hasn't changed, so no need to parse tree again and list of all suppressions is not initialized at all.
* There are issues outside TreeWalker, so Checker checks if there are any suppression for them. List is not initialized, thus `NullPointerException`.

I see several possible solutions:

1. Cache file should keep track of violations in all checks, not only  in `TreeWalker` (probably lots of work).
2. Always initialize list of suppressions to empty list, clear list if processing of new tree begins.
3. Put a null check and assume that if suppressions were not initialized by `TreeWalker` behaviour should be the same as for empty list of suppressions.

@romani, can you look into it and advise? On which level should it be tested? In which class? The bug is caused by combination of cache, TreeWalker and non-TreeWalker check.

**Comment by romani** [comment]
Yes , we have several problems (bad designs) with our cache file, some of them already registered as issues.

> Cache file should keep track of violations in all checks, not only in TreeWalker (probably lots of work).

YES.

> Always initialize list of suppressions to empty list, clear list if processing of new tree begins.

there should be a benefit from this in all scenarios.

> On which level should it be tested? In which class?

I have a rule to put a test method to Test class that combine all classes where changed during a fix.
I would put it to Checker (Treewalker, FileSetChecks, Cache).

**Comment by WonderCsabo** [comment]
OK, i just found this issue is partly covered in #407.

**Comment by Vladlis** [comment]
> Cache file should keep track of violations in all checks, not only in TreeWalker (probably lots of work).

Done in #2882 
> Always initialize list of suppressions to empty list, clear list if processing of new tree begins.

PR is opened - #3038

**Comment by romani** [comment]
Fix is merged

**Comment by WonderCsabo** [comment]
Thanks guys! Will this work with the current maven-checkstyle-plugin?

**Comment by Vladlis** [comment]
@WonderCsabo 
Sure, the main problem was fixed in Checkstyle 6.16.
